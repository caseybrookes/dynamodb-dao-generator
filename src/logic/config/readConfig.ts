import { DomainObject } from 'domain-objects';
import { DomainObjectMetadata, introspect } from 'domain-objects-metadata';

import { SupplementalQueryDeclaration } from '../../domain/objects/SupplementalQueryDeclaration';
import { UserInputError } from '../../utils/errors/UserInputError';
import { getDirOfPath } from '../../utils/filepaths/getDirOfPath';

/**
 * declares the path that should be introspected in order to find a superset of all of the domain objects we'll need metadata for
 */
export type DeclaredDomainObjectIntrospectionPaths = string[];

/**
 * the directories to output the generated dao to
 */
export interface DeclaredOutputDirectories {
  dao: string;
  terraform: string;
}

/**
 * the specification by which to generate each dao with
 */
export interface DeclaredDaoSpecification {
  domainObjectName: DomainObject<any>;
  supplementalIndexes?: SupplementalQueryDeclaration[];
}

export const readConfig = async ({
  configPath,
}: {
  configPath: string;
}): Promise<{
  metadatas: DomainObjectMetadata[];
  directories: DeclaredOutputDirectories;
  specifications: DeclaredDaoSpecification[];
}> => {
  const configDir = getDirOfPath(configPath);
  const getAbsolutePathFromRelativeToConfigPath = (relpath: string) =>
    `${configDir}/${relpath}`;

  try {
    const config = await import(configPath);
    if (!config.introspect)
      throw new UserInputError('no introspect key was exported from config');
    if (!config.directories)
      throw new UserInputError('no directories key was exported from config');
    if (!config.specifications)
      throw new UserInputError(
        'no specifications key was exported from config',
      );
    const metadatas: DomainObjectMetadata[] = (
      await Promise.all(
        config.introspect.map((path: string) =>
          introspect(getAbsolutePathFromRelativeToConfigPath(path)),
        ),
      )
    ).flat();
    return {
      metadatas,
      directories: {
        dao: getAbsolutePathFromRelativeToConfigPath(config.directories.dao),
        terraform: getAbsolutePathFromRelativeToConfigPath(
          config.directories.terraform,
        ),
      },
      specifications: config.specifications,
    };
  } catch (error) {
    console.error(error);
    throw error;
  }
};
