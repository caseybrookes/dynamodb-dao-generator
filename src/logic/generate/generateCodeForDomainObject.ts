import { DomainObjectMetadata } from 'domain-objects-metadata';

import { SupplementalQueryDeclaration } from '../../domain/objects/SupplementalQueryDeclaration';
import { defineTerraformResourceCodeForDomainObject } from '../define/terraform.resources/defineTerraformResourceCodeForDomainObject';
import { defineTypescriptDaoCodeForDomainObject } from '../define/typescript.dao/defineTypescriptDaoCodeForDomainObject';
import { outputGeneratedCodeFileToTargetDirectory } from '../output/outputGeneratedCodeFileToTargetDirectory';

export const generateCodeForDomainObject = async ({
  domainObjectMetadata,
  supplementalQueries,
  directories,
}: {
  domainObjectMetadata: DomainObjectMetadata;
  supplementalQueries: SupplementalQueryDeclaration[];
  directories: {
    terraform: string;
    dao: string;
  };
}) => {
  // define the terraform code
  const terraformCode = defineTerraformResourceCodeForDomainObject({
    domainObjectMetadata,
    supplementalQueries,
  });

  // define the typescript code
  const typescriptCode = defineTypescriptDaoCodeForDomainObject({
    domainObjectMetadata,
    supplementalQueries,
  });

  // output all of the files
  await outputGeneratedCodeFileToTargetDirectory({
    file: terraformCode,
    directory: directories.terraform,
  });
  await Promise.all(
    typescriptCode.map((file) =>
      outputGeneratedCodeFileToTargetDirectory({
        file,
        directory: directories.dao,
      }),
    ),
  );
};
