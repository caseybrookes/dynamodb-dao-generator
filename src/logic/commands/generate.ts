import chalk from 'chalk';

import { UnexpectedCodePathError } from '../../utils/errors/UnexpectedCodePathError';
import { readConfig } from '../config/readConfig';
import { generateCodeForDomainObject } from '../generate/generateCodeForDomainObject';

export const generate = async ({ configPath }: { configPath: string }) => {
  // read the declarations from config
  console.log(
    `${chalk.bold(
      '\n🔎 Loading domain objects:',
    )} using domain-objects-metadata...`,
  ); // tslint:disable-line no-console
  const config = await readConfig({ configPath });

  // output the sql-schema-generator entities
  console.log(
    `${chalk.bold(
      '\n🏗️️  Generating the dao:',
    )} methods, casters, terraform, and named exports...\n`,
  );
  await Promise.all(
    config.specifications.map((spec) => {
      // lookup the spec for this metadata
      const metadata = config.metadatas.find(
        (thisMetadata) => spec.domainObjectName === thisMetadata.name,
      );
      if (!metadata)
        throw new UnexpectedCodePathError(
          'could not find domain object metadata for specified domain object name. are you sure your introspection paths import or declare this domain object?',
          { domainObjectName: spec.domainObjectName },
        );

      // if it was identified though, generate the dao
      return generateCodeForDomainObject({
        domainObjectMetadata: metadata,
        directories: config.directories,
        supplementalQueries: spec.supplementalIndexes ?? [],
      });
    }),
  );
};
