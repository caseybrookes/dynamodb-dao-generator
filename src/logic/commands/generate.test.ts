import { DomainObjectMetadata } from 'domain-objects-metadata';

import { UnexpectedCodePathError } from '../../utils/errors/UnexpectedCodePathError';
import { DeclaredDaoSpecification, readConfig } from '../config/readConfig';
import { generateCodeForDomainObject } from '../generate/generateCodeForDomainObject';
import { generate } from './generate';

jest.mock('../config/readConfig');
const readConfigMock = readConfig as jest.Mock;

jest.mock('../generate/generateCodeForDomainObject');
const generateCodeForDomainObjectMock =
  generateCodeForDomainObject as jest.Mock;

describe('generate', () => {
  it('should throw a helpful error if the introspected domain objects do not include one specified in the config', async () => {
    // mock a config with a specified dobj that was not instrospected
    const exampleConfig: {
      metadatas: DomainObjectMetadata[];
      specifications: DeclaredDaoSpecification[];
    } = {
      metadatas: [],
      specifications: [{ domainObjectName: 'Sensor', supplementalIndexes: [] }],
    };
    readConfigMock.mockResolvedValue(exampleConfig);

    // run the request
    try {
      await generate({ configPath: '__config_path__' });
      throw new UnexpectedCodePathError('should not reach here');
    } catch (error) {
      // prove that it threw a helpful error
      expect(error.message).toContain(
        'could not find domain object metadata for specified domain object name. are you sure your introspection paths import or declare this domain object?',
      );
      expect(error).toMatchSnapshot();
    }

    // prove that it never succeeded
    expect(generateCodeForDomainObjectMock).not.toHaveBeenCalled();
  });
});
