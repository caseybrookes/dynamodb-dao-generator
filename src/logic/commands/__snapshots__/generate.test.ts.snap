// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate should throw a helpful error if the introspected domain objects do not include one specified in the config 1`] = `
[Error: UnexpectedCodePath: could not find domain object metadata for specified domain object name. are you sure your introspection paths import or declare this domain object?

{"domainObjectName":"Sensor"}]
`;
